src/NetMQ/Core/Patterns/Gather.cs

```diff
@@ -0,0 +1,105 @@
+    GNU Lesser General Public License for more details.
+
+    You should have received a copy of the GNU Lesser General Public License
+    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+using System.Diagnostics;
+using JetBrains.Annotations;
+using NetMQ.Core.Patterns.Utils;
+
+namespace NetMQ.Core.Patterns
+{
+    internal sealed class Gather : SocketBase
+    {
+        /// <summary>
+        /// Fair queueing object for inbound pipes.
+        /// </summary>
+        private readonly FairQueueing m_fairQueueing;
+
+        public Gather([NotNull] Ctx parent, int threadId, int socketId)
+            : base(parent, threadId, socketId)
+        {
+            m_options.SocketType = ZmqSocketType.Gather;
+
+            m_fairQueueing = new FairQueueing();
+        }
+
+        /// <summary>
+        /// Register the pipe with this socket.
+        /// </summary>
+        /// <param name="pipe">the Pipe to attach</param>
+        /// <param name="icanhasall">not used</param>
+        protected override void XAttachPipe(Pipe pipe, bool icanhasall)
+        {
+            Debug.Assert(pipe != null);
+            m_fairQueueing.Attach(pipe);
+        }
+
+        /// <summary>
+        /// Indicate the given pipe as being ready for reading by this socket.
+        /// </summary>
+        /// <param name="pipe">the <c>Pipe</c> that is now becoming available for reading</param>
+        protected override void XReadActivated(Pipe pipe)
+        {
+            m_fairQueueing.Activated(pipe);
+        }
+
+        /// <summary>
+        /// This is an override of the abstract method that gets called to signal that the given pipe is to be removed from this socket.
+        /// </summary>
+        /// <param name="pipe">the Pipe that is being removed</param>
+        protected override void XTerminated(Pipe pipe)
+        {
+            m_fairQueueing.Terminated(pipe);
+        }
+
+        /// <summary>
+        /// Receive a message. The <c>Recv</c> method calls this lower-level method to do the actual receiving.
+        /// </summary>
+        /// <param name="msg">the <c>Msg</c> to receive the message into</param>
+        /// <returns><c>true</c> if the message was received successfully, <c>false</c> if there were no messages to receive</returns>
+        protected override bool XRecv(ref Msg msg)
+        {
+            bool received = m_fairQueueing.Recv(ref msg);
+
+            // Drop any messages with more flag
+            while (received && msg.HasMore) 
+            {
+                // drop all frames of the current multi-frame message
+                received = m_fairQueueing.Recv(ref msg);
+
+                while (received && msg.HasMore)
+                    received = m_fairQueueing.Recv(ref msg);
+
+                // get the new message
+                if (received)
+                    received = m_fairQueueing.Recv(ref msg);
+            }
```

Comment:
I can't really tell what this is doing, but it looks more complicated than it needs to be. Why is `Recv` called in four places?

Reply:
Because thread-safe sockets don't support multi-frame messages the socket is dropping here any such messages.

Before the loop you get a new message, if it has more, you drop all the frames, and then before the loop exit, get a new message. The loop will end when you will get a message which is not multiframe.